### PR TITLE
pinv: forward/backward AD which is Frechet-defined in a rank-preserving neighborhood.

### LINKED ISSUE #65911: `pinv` could be differential on a wider ranger of inputs
## 🐛 Bug
The paper
`The Differentiation of Pseudo-Inverses and Nonlinear Least Squares Problems Whose Variables Separate. Author(s): G. H. Golub and V. Pereyra. Source: SIAM Journal on Numerical Analysis, Vol. 10, No. 2 (Apr., 1973), pp. 413-432`
states that `pinv` is differentiable in a rank-preserving neighborhood.

However, given that `pinv` is implemented via the SVD, its backward suffers from all the same issue: no repeated eigenvalues are allowed.

## To Reproduce

Here is an example of a rank-preserving transformation that should be differentiable but it is not:

```python
In [1]: import torch

In [2]: x = torch.rand(30, 1, dtype=torch.double, requires_grad=True)

In [3]: y = torch.rand(30, 1, dtype=torch.double, requires_grad=True)                                                                                                                                                                                                                                                       

In [4]: torch.autograd.gradcheck(lambda a, b: torch.linalg.pinv(a @ b.t()), [x, y])
---------------------------------------------------------------------------
GradcheckError                            Traceback (most recent call last)
<ipython-input-4-5e7e12923389> in <module>()
----> 1 torch.autograd.gradcheck(lambda a, b: torch.linalg.pinv(a @ b.t()), [x, y])

~/git/Quansight/pytorch/torch/autograd/gradcheck.py in gradcheck(func, inputs, eps, atol, rtol, raise_exception, check_sparse_nnz, nondet_tol, check_undefined_grad, check_grad_dtypes, check_batched_grad, check_forward_ad, check_backward_ad, fast_mode)                                                                 
   1271             return False
   1272     else:
-> 1273         return _gradcheck_helper(**args)
   1274
   1275

~/git/Quansight/pytorch/torch/autograd/gradcheck.py in _gradcheck_helper(func, inputs, eps, atol, rtol, check_sparse_nnz, nondet_tol, check_undefined_grad, check_grad_dtypes, check_batched_grad, check_forward_ad, check_backward_ad, fast_mode)                                                                          
   1286     _gradcheck_real_imag(gradcheck_fn, func, func_out, tupled_inputs, outputs, eps,
   1287                          rtol, atol, check_grad_dtypes, check_forward_ad=check_forward_ad,
-> 1288                          check_backward_ad=check_backward_ad, nondet_tol=nondet_tol)
   1289     # Short circuit because remaining tests rely on backward AD to be implemented
   1290     if not check_backward_ad:

~/git/Quansight/pytorch/torch/autograd/gradcheck.py in _gradcheck_real_imag(gradcheck_fn, func, func_out, tupled_inputs, outputs, eps, rtol, atol, check_grad_dtypes, check_forward_ad, check_backward_ad, nondet_tol)                                                                                                      
    946         else:
    947             gradcheck_fn(func, func_out, tupled_inputs, outputs, eps,
--> 948                          rtol, atol, check_grad_dtypes, nondet_tol)
    949
    950     if check_forward_ad:

~/git/Quansight/pytorch/torch/autograd/gradcheck.py in _slow_gradcheck(func, func_out, tupled_inputs, outputs, eps, rtol, atol, check_grad_dtypes, nondet_tol, use_forward_ad, complex_indices, test_imag)                                                                                                                  
    994             for j, (a, n) in enumerate(zip(analytical, numerical[i])):
    995                 if not _allclose_with_type_promotion(a, n.to(a.device), rtol, atol):
--> 996                     raise GradcheckError(_get_notallclose_msg(a, n, i, j, complex_indices, test_imag))
    997
    998     return True

GradcheckError: Jacobian mismatch for output 0 with respect to input 0,
numerical:tensor([[ 9.9343e-04, -1.7439e-05, -5.0706e-06,  ..., -4.6110e-05,
         -9.7824e-05, -5.8797e-05],
        [-1.7439e-05,  9.1318e-04, -2.4387e-05,  ..., -2.2177e-04,
         -4.7049e-04, -2.8279e-04],
        [-5.0706e-06, -2.4387e-05,  9.8997e-04,  ..., -6.4482e-05,
         -1.3680e-04, -8.2224e-05],
        ...,
        [-1.2011e-05, -5.7770e-05, -1.6797e-05,  ...,  3.6748e-03,
         -3.2406e-04, -1.9478e-04],
        [-2.5482e-05, -1.2256e-04, -3.5635e-05,  ..., -3.2406e-04,
          3.1401e-03, -4.1322e-04],
        [-1.5316e-05, -7.3665e-05, -2.1419e-05,  ..., -1.9478e-04,
         -4.1322e-04,  3.5792e-03]], dtype=torch.float64)
analytical:tensor([[nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        ...,
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan]], dtype=torch.float64)


```

## Additional context

I still have to read the paper in depth to see whether we can fix the backward once and for all.
Still though, even without the paper we could still use the SVD in backward where we only consider the
first rank of the input subspaces. This way we still suffer from the SVD backward issue (no repeated singular values),
but at least we can handle arbitrary rank if non-zero singular values are distinct.

Once we have explicit backward for `pinv`, it could be used for `lstsq` to make it differentiable.

### PR DESCRIPTION
**Pull request by nikitaved**
Fixes https://github.com/pytorch/pytorch/issues/65911. Also enables complex support/tests for `linalg_pinv` in OpInfo.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @jianyuh @mruberry @walterddr @IvanYashchuk @xwang233

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -8908,14 +8931,36 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
...
+           # to avoid any rank changes caused by the perturbations in the gradcheck
+           op=lambda a, b: torch.linalg.pinv(a @ b.transpose(-1, -2)),
+           dtypes=floating_and_complex_types(),
+           supports_out=False,
+           check_batched_grad=False,
+           check_batched_gradgrad=False,
+           supports_forward_ad=True,
+           sample_inputs_func=sample_inputs_linalg_pinv_singular,
+           # Only large tensors show issues with implicit backward used prior to
+           # explicit backward implementation.
```
Comment (nikitaved) [inline torch/testing/_internal/common_methods_invocations.py:8951]: a note: large tensors of low rank. In my environment I had to create a 1-rank 30x30 matrix to see issues with repeated zeros in the backward of SVD.

**Comment by ezyang** [comment]
not sure appropriate FB reviewer has been tagged yet

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -2102,6 +2102,29 @@ def sample_inputs_linalg_invertible(op_info, device, dtype, requires_grad=False,
...
+    batches = [(), (0, ), (2, ), (1, 1)]
+    # the size of at least 30 is required to cause failures for the previous implicit implementation
+    # of the pinv's backward method, albeit it is slow.
+    size = [0, 3, 30, 50, 80]
+
+    def generate_samples():
+        for batch, m, n in product(batches, size, size):
+            for k in range(min(3, min(m, n))):
+                # Note that by making the columns of `a` and `b` orthonormal we make sure
+                # that the product matrix `a @ b.t()` has condition number 1.
```
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:2121]: Nice! This saves us a lot of pain in future debugging.

Now, this note is slightly incorrect. The resulting matrix will have singular values 0 and 1, so the condition number will be infinite! Perhaps you mean that it has condition number 1 when restricted to its image?

**Review by lezcano — APPROVED** [review]
Nice! Thanks both for finding the slick formula for this backward and the rather compact implementation!

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:2121]: Yes, exactly in the image, correct.

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -2102,6 +2102,29 @@ def sample_inputs_linalg_invertible(op_info, device, dtype, requires_grad=False,
...
+    """
+    This function produces factors `a` and `b` to generate inputs of the form `a @ b.t()` to
+    test the backward method of `linalg_pinv`. That way we always preserve the rank of the
+    input no matter the perturbations applied to it by the gradcheck.
+    Note that `pinv` is Frechet-differentiable in a rank-preserving neighborhood.
+    """
+    batches = [(), (0, ), (2, ), (1, 1)]
+    # the size of at least 30 is required to cause failures for the previous implicit implementation
+    # of the pinv's backward method, albeit it is slow.
+    size = [0, 3, 30, 50, 80]
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:2115]: Should we have only one above 30 here to reduce time? Like only 50?

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -8908,14 +8931,36 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
...
+           op=lambda a, b: torch.linalg.pinv(a @ b.transpose(-1, -2)),
+           dtypes=floating_and_complex_types(),
+           supports_out=False,
+           check_batched_grad=False,
+           check_batched_gradgrad=False,
+           supports_forward_ad=True,
+           sample_inputs_func=sample_inputs_linalg_pinv_singular,
+           # Only large tensors show issues with implicit backward used prior to
+           # explicit backward implementation.
+           decorators=[slowTest, skipCUDAIfNoMagmaAndNoCusolver, skipCUDAIfRocm, skipCPUIfNoLapack],
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:8952]: Is the slowTest decorator working as expected here?

Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:8952]: Yes!

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:2115]: Wanted to really push the boundaries to see whether all envs will handle this.

**Comment by mruberry** [comment]
Cool! Do you have before/after perf numbers for the autograd, @nikitaved?

**Review by albanD — APPROVED** [review]
Looks good. Can you fix the last lint (EDIT: Ho it looks like the job itself failed...)and I'll merge this.

**Pull request closed without merging**